In [1]:
# Importing relevant libraries for data manipulation
import numpy as np
import pandas as pd 

In [2]:
# Load dataset from source directory into a dataframe
df = pd.read_csv(r"C:\Users\chris\OneDrive\Documents\paschal\MY PORTFOLIO PROJECTS\CUSTOMER CHURN\Churn_Modelling.csv")

### DATA CLEANING 


In [3]:
# Display data summary for familiarization
df.head(10)


RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   
5          6    15574012       Chu          645     Spain    Male   44   
6          7    15592531  Bartlett          822    France    Male   50   
7          8    15656148    Obinna          376   Germany  Female   29   
8          9    15792365        He          501    France    Male   44   
9         10    15592389        H?          684    France    Male   27   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   
5       8  113755.78              2          1               0   
6       7       0.00              2          1               1   
7       4  115046.74              4          1               0   
8       4  142051.07              2          0               1   
9       2  134603.88              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0  
5        149756.71       1  
6         10062.80       0  
7        119346.88       1  
8         74940.50       0  
9         71725.73       0

In [4]:
# Data exploration
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


In [5]:
# Display columns
print(df.columns)

Index(['RowNumber', 'CustomerId', 'Surname', 'CreditScore', 'Geography',
       'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard',
       'IsActiveMember', 'EstimatedSalary', 'Exited'],
      dtype='object')


In [6]:
# Descriptive statistics
df.describe()

RowNumber    CustomerId   CreditScore           Age        Tenure  \
count  10000.00000  1.000000e+04  10000.000000  10000.000000  10000.000000   
mean    5000.50000  1.569094e+07    650.528800     38.921800      5.012800   
std     2886.89568  7.193619e+04     96.653299     10.487806      2.892174   
min        1.00000  1.556570e+07    350.000000     18.000000      0.000000   
25%     2500.75000  1.562853e+07    584.000000     32.000000      3.000000   
50%     5000.50000  1.569074e+07    652.000000     37.000000      5.000000   
75%     7500.25000  1.575323e+07    718.000000     44.000000      7.000000   
max    10000.00000  1.581569e+07    850.000000     92.000000     10.000000   

             Balance  NumOfProducts    HasCrCard  IsActiveMember  \
count   10000.000000   10000.000000  10000.00000    10000.000000   
mean    76485.889288       1.530200      0.70550        0.515100   
std     62397.405202       0.581654      0.45584        0.499797   
min         0.000000       1.000000      0.00000        0.000000   
25%         0.000000       1.000000      0.00000        0.000000   
50%     97198.540000       1.000000      1.00000        1.000000   
75%    127644.240000       2.000000      1.00000        1.000000   
max    250898.090000       4.000000      1.00000        1.000000   

       EstimatedSalary        Exited  
count     10000.000000  10000.000000  
mean     100090.239881      0.203700  
std       57510.492818      0.402769  
min          11.580000      0.000000  
25%       51002.110000      0.000000  
50%      100193.915000      0.000000  
75%      149388.247500      0.000000  
max      199992.480000      1.000000

In [7]:
# Checking for NULLs 
df.isnull().sum()

RowNumber          0
CustomerId         0
Surname            0
CreditScore        0
Geography          0
Gender             0
Age                0
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          0
IsActiveMember     0
EstimatedSalary    0
Exited             0
dtype: int64

In [8]:
# Remove duplicates if any
df = df.drop_duplicates()

In [9]:
# Verify changes 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


### DATA PREPROCESSING and FEATURE ENGINEERING

In [10]:
# To change the values in HasCrCard, IsActiveMember from integer to object datatype
# Create a function to replace 1 with 'yes' and 0 with 'no'

def map_values(column):
    return column.map({1: 'Yes', 0: 'No'})

# Applying the mapping function to 'IsActiveMember' and 'HasCrCard' columns
df['HasCrCard'] = map_values(df['HasCrCard'])
df['IsActiveMember'] = map_values(df['IsActiveMember'])

# Check updated dataframe
print(df)


      RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts HasCrCard IsActiveMember

In [11]:
# Visalize altered column 
print(df[['HasCrCard', 'IsActiveMember']])

     HasCrCard IsActiveMember
0          Yes            Yes
1           No            Yes
2          Yes             No
3           No             No
4          Yes            Yes
...        ...            ...
9995       Yes             No
9996       Yes            Yes
9997        No            Yes
9998       Yes             No
9999       Yes             No

[10000 rows x 2 columns]


In [12]:
# To change the values in Exited columns from a numeric to object datatype
 
# Create a function to map 1 to 'Churned' and 0 to 'Retained'
def map_values_exited(column):
    return column.map({1: 'Churned', 0: 'Retained'})

# Applying the mapping function to 'Exited columns
df['Exited'] = map_values_exited(df['Exited'])

# Check updated dataframe
print(df)



      RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts HasCrCard IsActiveMember

In [13]:
# Visalize altered column 
print(df['Exited'])

0        Churned
1       Retained
2        Churned
3       Retained
4       Retained
          ...   
9995    Retained
9996    Retained
9997     Churned
9998     Churned
9999    Retained
Name: Exited, Length: 10000, dtype: object


In [14]:
# Verify changes
df.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts HasCrCard IsActiveMember  EstimatedSalary  \
0       2       0.00              1       Yes            Yes        101348.88   
1       1   83807.86              1        No            Yes        112542.58   
2       8  159660.80              3       Yes             No        113931.57   
3       1       0.00              2        No             No         93826.63   
4       2  125510.82              1       Yes            Yes         79084.10   

     Exited  
0   Churned  
1  Retained  
2   Churned  
3  Retained  
4  Retained

### EXPLORATORY DATA ANALYSIS 

In [15]:
# Distribution of customer churn in the Exited column
exited_distribution = df['Exited'].value_counts().reset_index()
exited_distribution.columns = ['Exited? (0 = Retained, 1 = Churned)', 'Count']

# Adding proportions
exited_distribution['Proportion (%)'] = (exited_distribution['Count'] / df.shape[0]) * 100

# Display the distribution
print("Distribution of Customer churn:")
print(exited_distribution)

Distribution of Customer churn:
  Exited? (0 = Retained, 1 = Churned)  Count  Proportion (%)
0                            Retained   7963           79.63
1                             Churned   2037           20.37


### Initial tentative observation
The churn rate which is a key business metric is pegged at `20.37%`.

This confirmed that approx 20% of customers churned indicating that `1 in 5 customers` churned over the past `10` Tenures cycles

In [16]:
# Grouping Age and Balance by Exited

# Identifying the average Age and Balance of the customers
age_balance_vs_exited = df.groupby('Exited')[['Age', 'EstimatedSalary']].mean()
print("Mean Age and Balance by Exited:")
print(age_balance_vs_exited)

Mean Age and Balance by Exited:
                Age  EstimatedSalary
Exited                              
Churned   44.837997    101465.677531
Retained  37.408389     99738.391772


### Initial tentative observation

On an average the Churned customers are seen to be older, approx `45 years`, than the retained customers at approx `37 years` which suggests that older customers are more likely to churn, potentially due to changing needs, dissatisfaction, or financial constraints as they age.

But contrary to what might be expected these churned customers earn slightly more on an average i.e `$101,465.68` as against `$99,738.39` for customers that were retained  indicating that higher-income customers might have access to more competitive offers or are more likely to switch to services that better meet their expectations which indicating income levesl as a key influence in their likelihood of switching providers.

In [17]:
# Separate the Age of customers based on churn status
churned_ages = df[df['Exited'] == 1]['Age']
retained_ages = df[df['Exited'] == 0]['Age']
print(churned_ages)
print(retained_ages)

Series([], Name: Age, dtype: int64)
Series([], Name: Age, dtype: int64)


In [18]:
# Creating a new column called 'age_group' to aggregate customer ages into specified categories.
def categorize_age(age):
    if 18 <= age <= 35:
        return 'Young Adults(18-35)' # For ages 18 - 35
    elif 36 <= age <= 60:
        return 'Middle Aged(36-60)' # For ages 36 - 60
    elif age >= 61:
        return 'Seniors(61 and above)' # For ages above 60
    else:
        return 'Out of Range'  # For ages less than 18 

# Apply the function to the "age" column to create a new "age_group" column
df['age_group'] = df['Age'].apply(categorize_age)

# Display results
print(df)

      RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts HasCrCard IsActiveMember

In [19]:
# Age distribution for all customers

# Get the value counts for the categorized age groups
age_group_counts = df['age_group'].value_counts().reset_index()

# Rename the columns for clarity
age_group_counts.columns = ['Age Group', 'Count']

# Add proportions for each age group
age_group_counts['Proportion (%)'] = (age_group_counts['Count'] / df.shape[0]) * 100

# Display the results
print("Categorized Age distribution for all customers:")
print(age_group_counts)

Categorized Age distribution for all customers:
               Age Group  Count  Proportion (%)
0     Middle Aged(36-60)   5383           53.83
1    Young Adults(18-35)   4153           41.53
2  Seniors(61 and above)    464            4.64


### Initial tentative observation

The demographic of customers were categorized into three main age groups highlighted by different stages in adulthood

`Young Adults (18-35)`: Represents the youngest segment of the customer base.

`Middle-Aged (36-60)`: Covers the majority of working-age adults.

`Seniors (Above 60)`: Represents the older demographic segment.

Middle Aged Adults accounted for more than half of total customer base at `53.8%` while Young Adults followed closely at `41.5%`; the Seniors were significamtly smaller in size at just `4.7%`. 

This clearly suggests that Middle Aged Adults forms the core demographic for the company showing that the company’s products/services may appeal most strongly to this group.

The significant presence of middle-aged customers underscores the importance of maintaining loyalty in this group, while the smaller senior segment highlights an opportunity to grow in an underserved market.

In [20]:
# Isolating the customer age group with the highest churn 

# Group by age_group and Exited, then calculate the churn count
churn_counts = df.groupby(['age_group', 'Exited']).size().unstack(fill_value=0)

# Add total customers column
churn_counts['Total_Customers'] = churn_counts['Churned'] + churn_counts['Retained']

# Calculate churn and retention rates for each age group
churn_counts['Churn_Rate'] = (churn_counts['Churned'] / churn_counts['Total_Customers'] * 100).round(2)
churn_counts['Retention_Rate'] = (churn_counts['Retained'] / churn_counts['Total_Customers'] * 100).round(2)

# Calculate percentage of total customers in each age group
total_customers = churn_counts['Total_Customers'].sum()
churn_counts['Percent_of_Total'] = (churn_counts['Total_Customers'] / total_customers * 100).round(2)

# Identify the age groups with highest churn and retention rates
highest_churn_group = churn_counts['Churn_Rate'].idxmax()
highest_retention_group = churn_counts['Retention_Rate'].idxmax()

# Display the results
print("Customer Age Group vs churn Analysis:")
print("\nDetailed Counts and Rates:")
print("-------------------------")
print(churn_counts[['Churned', 'Retained', 'Total_Customers', 'Percent_of_Total', 'Churn_Rate', 'Retention_Rate']])
print("\nSummary:")
print(f"Total Customers: {total_customers}")
print(f"Age group with highest churn rate: {highest_churn_group} ({churn_counts.loc[highest_churn_group, 'Churn_Rate']}%)")
print(f"Age group with highest retention rate: {highest_retention_group} ({churn_counts.loc[highest_retention_group, 'Retention_Rate']}%)")

Customer Age Group vs churn Analysis:

Detailed Counts and Rates:
-------------------------
Exited                 Churned  Retained  Total_Customers  Percent_of_Total  \
age_group                                                                     
Middle Aged(36-60)        1575      3808             5383             53.83   
Seniors(61 and above)      115       349              464              4.64   
Young Adults(18-35)        347      3806             4153             41.53   

Exited                 Churn_Rate  Retention_Rate  
age_group                                          
Middle Aged(36-60)          29.26           70.74  
Seniors(61 and above)       24.78           75.22  
Young Adults(18-35)          8.36           91.64  

Summary:
Total Customers: 10000
Age group with highest churn rate: Middle Aged(36-60) (29.26%)
Age group with highest retention rate: Young Adults(18-35) (91.64%)


### Initial tentative observation

Customers were categorized based on their age groups i.e Young Adults, Middle Aged and Senior Adults.

Middle Aged Adults had the highest churn rate at `29%` nearly three times that of the young adults at `8%` but just slightly above the Seniors at `25%` which points to the fact that customers in this category might be more demanding and financially affluent, leading to dissatisfaction and eventual churn if services don't meet expectations. 

It is worth noting that Seniors still had a relatively high churn rate just slightly below Middle Aged adults; most likey due to their assumed financial limitations, retirement, or shifting preferences, further analysis would be required to confirm this.

Young Adults were the most loyal by a distance `92% retained`; They represent an opportunity for growth through targeted marketing and retention strategies.





In [21]:
# Categorizing Salaries of all customers to discover the distribution of their earnings

# Function to categorize EstimatedSalary
def categorize_salary(salary):
    if 11.580 <= salary <= 66671.88:
        return 'Low salary'  # $11.58 - $66,671.88
    elif 66671.88 < salary <= 133332.18:
        return 'Medium salary'  # $66,671.88 - $133,332.18
    elif 133332.18 < salary <= 199992.48:
        return 'High salary'  # $133,332.18 - $199,992.48
    else:
        return 'Out of Range'  # Just in case there are any unexpected values

# Apply the function to create a new column
df['salary_category'] = df['EstimatedSalary'].apply(categorize_salary)

# Display the first few rows of the dataset to confirm the new column
print("First few rows with the salary category:")
print(df[['EstimatedSalary', 'salary_category']].head())

# Count the number of customers in each category
salary_category_counts = df['salary_category'].value_counts()

# Convert counts to a DataFrame and calculate proportions
salary_category_df = salary_category_counts.to_frame(name='Count')
salary_category_df['Proportion (%)'] = (salary_category_df['Count'] / df.shape[0]) * 100

print("\nDistribution of salary categories with proportions:")
print(salary_category_df)


First few rows with the salary category:
   EstimatedSalary salary_category
0        101348.88   Medium salary
1        112542.58   Medium salary
2        113931.57   Medium salary
3         93826.63   Medium salary
4         79084.10   Medium salary

Distribution of salary categories with proportions:
                 Count  Proportion (%)
salary_category                       
Medium salary     3383           33.83
Low salary        3314           33.14
High salary       3303           33.03


### Initial tentative observation

The Estimated salary of all customers was categorized based on level of their earnings:

`Low Salary (between $11.58 - $66,671.88)`: Customers in this category are likely in entry-level jobs, part-time employment, or lower-paying industries.

`Medium Salary ($66,671.88 - $133,332.18)`: Customers that fall into this category, are more likely financially stable and composed of middle-aged individuals in mid-level positions or professionals with consistent earnings. They are usually able to access a broader range of financial products, such as savings plans, loans, or moderate investment options.

`High Salary ($133,332.18 - $199,992.48)`: Customers in this bracket make the highest earnings, potentially senior professionals, business owners, or retirees with significant financial resources. They are likely to seek premium financial services or luxury-focused services.


There are very slight differences between the number of customers in these categories with those in the Medium class being the most predominant at `3383 (33.83% of the customer base)` followed closely by Low salary class at `3314 (33.14% of the customer base)`, with the Highest earners being the fewest at a little over `3000 customers (33.03%)`.

There is an even distribution amongst these income groups meaning all are well represented overall. 
The "Medium" salary category has the highest representation, indicating that most customers fall within this income bracket.

In [22]:
# Age groups vs. salary categories analysis

# Grouping data by age groups and salary categories
age_salary_analysis = df.groupby(['age_group', 'salary_category']).size().unstack(fill_value=0)

# Adding a total count per age group
age_salary_analysis['Total'] = age_salary_analysis.sum(axis=1)

# Calculating proportions for each salary category within each age group
age_salary_analysis['Low salary (%)'] = (age_salary_analysis['Low salary'] / age_salary_analysis['Total']) * 100
age_salary_analysis['Medium salary (%)'] = (age_salary_analysis['Medium salary'] / age_salary_analysis['Total']) * 100
age_salary_analysis['High salary (%)'] = (age_salary_analysis['High salary'] / age_salary_analysis['Total']) * 100

# Identifying the age group with the highest count in each salary category
max_low_group = age_salary_analysis['Low salary'].idxmax()
max_medium_group = age_salary_analysis['Medium salary'].idxmax()
max_high_group = age_salary_analysis['High salary'].idxmax()

# Display the detailed table
print("Age groups vs. salary categories analysis:")
print(age_salary_analysis)

print("\nObservations:")
print(f"The age group with the most low salary earners is: {max_low_group}")
print(f"The age group with the most medium salary earners is: {max_medium_group}")
print(f"The age group with the most high salary earners is: {max_high_group}")

Age groups vs. salary categories analysis:
salary_category        High salary  Low salary  Medium salary  Total  \
age_group                                                              
Middle Aged(36-60)            1768        1728           1887   5383   
Seniors(61 and above)          144         170            150    464   
Young Adults(18-35)           1391        1416           1346   4153   

salary_category        Low salary (%)  Medium salary (%)  High salary (%)  
age_group                                                                  
Middle Aged(36-60)          32.101059          35.054802        32.844139  
Seniors(61 and above)       36.637931          32.327586        31.034483  
Young Adults(18-35)         34.095834          32.410306        33.493860  

Observations:
The age group with the most low salary earners is: Middle Aged(36-60)
The age group with the most medium salary earners is: Middle Aged(36-60)
The age group with the most high salary earners is: Middle

### Initial tentative observation

The dataset was further categorized to extrapolate the age group with the highest earning potential based on their Estimated salaries

`Middle Aged customers` dominated in all earning categories ranking the highest overall. Having a substantial number of customers in the `Medium salary category at 35%`, makes it the age group with the highest count of Mid-income earners.
It also maintains a healthy proportion in the `High salarycategory (33%)` while having the lowest percentage in the `Low salary category (32%)`. This observation reflects that middle-aged individuals are at the peak of their earning potential, often holding senior positions in their careers or managing well-established businesses.

Surprisingly `Young adults` came a close second in all categories which on deeper analysis can be attributed to their large numbers compared to seniors; they comprise a significant portion of the dataset and were predominantly but unsurprsingly found in the `Low salary category (34%)`; being younger individuals, they are often early in their careers and are still building their earning potential.
A relatively smaller percentage of customers in this group fall into the `High salary (33%)` bracket though slightly larger than the percentage in the Medium salary category, nonetheless still indicative that the likelihood of young adults being top earners is low compared to other age groups.

`Seniors`, while fewer in total population, surprisingly have a higher proportion of its members in the `Low salary category (37%)`relative to their total population with their least proprtion in the `High salary category (33%)`.
This suggests that although fewer seniors are part of the dataset, those who tend to have financial limitations, retirement and relatively poor financial stability are relatively high.



In [23]:
# Customer Salary Category vs. Churn Analysis

# Group by salary_category and Exited, then calculate the churn count
salary_churn_counts = df.groupby(['salary_category', 'Exited']).size().unstack(fill_value=0)

# Add total customers column
salary_churn_counts['Total_Customers'] = salary_churn_counts['Churned'] + salary_churn_counts['Retained']

# Calculate churn and retention rates for each salary category
salary_churn_counts['Churn_Rate'] = (salary_churn_counts['Churned'] / salary_churn_counts['Total_Customers'] * 100).round(2)
salary_churn_counts['Retention_Rate'] = (salary_churn_counts['Retained'] / salary_churn_counts['Total_Customers'] * 100).round(2)

# Calculate percentage of total customers in each salary category
total_customers = salary_churn_counts['Total_Customers'].sum()
salary_churn_counts['Percent_of_Total'] = (salary_churn_counts['Total_Customers'] / total_customers * 100).round(2)

# Identify the salary categories with highest churn and retention rates
highest_churn_salary = salary_churn_counts['Churn_Rate'].idxmax()
highest_retention_salary = salary_churn_counts['Retention_Rate'].idxmax()

# Display the results
print("Customer Salary Category vs. Churn Analysis:")
print("\nDetailed Counts and Rates:")
print("-------------------------")
print(salary_churn_counts[['Churned', 'Retained', 'Total_Customers', 'Percent_of_Total', 'Churn_Rate', 'Retention_Rate']])
print("\nSummary:")
print(f"Total Customers: {total_customers}")
print(f"Salary category with highest churn rate: {highest_churn_salary} ({salary_churn_counts.loc[highest_churn_salary, 'Churn_Rate']}%)")
print(f"Salary category with highest retention rate: {highest_retention_salary} ({salary_churn_counts.loc[highest_retention_salary, 'Retention_Rate']}%)")


Customer Salary Category vs. Churn Analysis:

Detailed Counts and Rates:
-------------------------
Exited           Churned  Retained  Total_Customers  Percent_of_Total  \
salary_category                                                         
High salary          701      2602             3303             33.03   
Low salary           665      2649             3314             33.14   
Medium salary        671      2712             3383             33.83   

Exited           Churn_Rate  Retention_Rate  
salary_category                              
High salary           21.22           78.78  
Low salary            20.07           79.93  
Medium salary         19.83           80.17  

Summary:
Total Customers: 10000
Salary category with highest churn rate: High salary (21.22%)
Salary category with highest retention rate: Medium salary (80.17%)


   ### Initial tentative observation

Customers that earn `High salaries ($133,332.18 - $199,992.48)` have the highest churn rates at `21%` and lowest retention rates at `79%` is an indicator that they churn likely to seek premium financial services or luxury-focused services elsewhere. F
urther analysis and correlation would need to be made to augment this tentative conclusion.

Medium and Low salary earners had the highest retenetion at approx `80%` signifying overall satisfactory customer satisfaction at these income brackets



In [24]:
# Customer Gender vs. Churn Analysis 

# Grouping data by gender and Exited status
gender_churn_analysis = df.groupby(['Gender', 'Exited']).size().unstack(fill_value=0)

# Adding total customers column
gender_churn_analysis['Total_Customers'] = gender_churn_analysis['Churned'] + gender_churn_analysis['Retained']

# Calculate churn and retention rates for each gender
gender_churn_analysis['Churn_Rate'] = (gender_churn_analysis['Churned'] / gender_churn_analysis['Total_Customers'] * 100).round(2)
gender_churn_analysis['Retention_Rate'] = (gender_churn_analysis['Retained'] / gender_churn_analysis['Total_Customers'] * 100).round(2)

# Calculate percentage of total customers in each gender
total_customers = gender_churn_analysis['Total_Customers'].sum()
gender_churn_analysis['Percent_of_Total'] = (gender_churn_analysis['Total_Customers'] / total_customers * 100).round(2)

# Identify the gender with highest churn and retention rates
highest_churn_gender = gender_churn_analysis['Churn_Rate'].idxmax()
highest_retention_gender = gender_churn_analysis['Retention_Rate'].idxmax()

# Display the results
print("Customer Gender vs. Churn Analysis:")
print("\nDetailed Counts and Rates:")
print("-------------------------")
print(gender_churn_analysis[['Churned', 'Retained', 'Total_Customers', 'Percent_of_Total', 'Churn_Rate', 'Retention_Rate']])
print("\nSummary:")
print(f"Total Customers: {total_customers}")
print(f"Gender with highest churn rate: {highest_churn_gender} ({gender_churn_analysis.loc[highest_churn_gender, 'Churn_Rate']}%)")
print(f"Gender with highest retention rate: {highest_retention_gender} ({gender_churn_analysis.loc[highest_retention_gender, 'Retention_Rate']}%)")



Customer Gender vs. Churn Analysis:

Detailed Counts and Rates:
-------------------------
Exited  Churned  Retained  Total_Customers  Percent_of_Total  Churn_Rate  \
Gender                                                                     
Female     1139      3404             4543             45.43       25.07   
Male        898      4559             5457             54.57       16.46   

Exited  Retention_Rate  
Gender                  
Female           74.93  
Male             83.54  

Summary:
Total Customers: 10000
Gender with highest churn rate: Female (25.07%)
Gender with highest retention rate: Male (83.54%)


### Initial tentative observation

There are more male customers than female overall i.e. `55%` to `45%` 

Female customers are `25%` more likely to churn than their male counterparts who were `83%` more likely to be retained.
This is a key metric to fully understanding the factors contributing to churn 



In [25]:
# Customer Gender vs. Salary Category Analysis

# Grouping data by gender and salary categories
gender_salary_analysis = df.groupby(['Gender', 'salary_category']).size().unstack(fill_value=0)

# Adding total customers column
gender_salary_analysis['Total_Customers'] = gender_salary_analysis.sum(axis=1)

# Calculate percentage distribution of salary categories within each gender
for category in ['Low salary', 'Medium salary', 'High salary']:
    gender_salary_analysis[f'{category} (%)'] = (gender_salary_analysis[category] / gender_salary_analysis['Total_Customers'] * 100).round(2)

# Calculate percentage of total customers by gender
total_customers = gender_salary_analysis['Total_Customers'].sum()
gender_salary_analysis['Percent_of_Total'] = (gender_salary_analysis['Total_Customers'] / total_customers * 100).round(2)

# Identify the predominant salary category for each gender
predominant_categories = gender_salary_analysis[['Low salary', 'Medium salary', 'High salary']].idxmax(axis=1)

# Display the results
print("Customer Gender vs. Salary Category Analysis:")
print("\nDetailed Counts and Percentages:")
print("-------------------------")
print(gender_salary_analysis[['Low salary', 'Medium salary', 'High salary', 
                             'Total_Customers', 'Percent_of_Total',
                             'Low salary (%)', 'Medium salary (%)', 'High salary (%)']])
print("\nSummary:")
print(f"Total Customers: {total_customers}")
print("\nPredominant salary category by gender:")
for gender in predominant_categories.index:
    print(f"{gender}: {predominant_categories[gender]} "
          f"({gender_salary_analysis.loc[gender, predominant_categories[gender]]} customers, "
          f"{gender_salary_analysis.loc[gender, f'{predominant_categories[gender]} (%)']}%)")



Customer Gender vs. Salary Category Analysis:

Detailed Counts and Percentages:
-------------------------
salary_category  Low salary  Medium salary  High salary  Total_Customers  \
Gender                                                                     
Female                 1486           1557         1500             4543   
Male                   1828           1826         1803             5457   

salary_category  Percent_of_Total  Low salary (%)  Medium salary (%)  \
Gender                                                                 
Female                      45.43           32.71              34.27   
Male                        54.57           33.50              33.46   

salary_category  High salary (%)  
Gender                            
Female                     33.02  
Male                       33.04  

Summary:
Total Customers: 10000

Predominant salary category by gender:
Female: Medium salary (1557 customers, 34.27%)
Male: Low salary (1828 customers, 33.5%)

### Initial tentative observation
There are more male customers than female overall i.e. `55%` to `45%` 

Men dominated the all salary categories across board by their share numbers falling into the different salary category buckets (over `1800`men in all buckets against the womens `1500` average).

Nonethelss by the percentage of their respective poulations falling in these buckets, Female customers had the higher percentage of medium salary earners at `34%` of their population relative to their male counterparts (`33%`) who then ranked higher in the low salary earners with `34%` of their relative population in that bucket. 

Interestingly they have an approximate equal standing in the high salary category at `33%` each, but women leading ever so sligtly with a `0.02 difference`.



In [26]:
# Customer Gender vs. Age Group Analysis

# Grouping data by gender and age groups
gender_age_analysis = df.groupby(['Gender', 'age_group']).size().unstack(fill_value=0)

# Adding total customers column
gender_age_analysis['Total_Customers'] = gender_age_analysis.sum(axis=1)

# Calculate percentage distribution of age groups within each gender
for age_group in ['Young Adults(18-35)', 'Middle Aged(36-60)', 'Seniors(61 and above)']:
    gender_age_analysis[f'{age_group} (%)'] = (gender_age_analysis[age_group] / gender_age_analysis['Total_Customers'] * 100).round(2)

# Calculate percentage of total customers by gender
total_customers = gender_age_analysis['Total_Customers'].sum()
gender_age_analysis['Percent_of_Total'] = (gender_age_analysis['Total_Customers'] / total_customers * 100).round(2)

# Identify the predominant age group for each gender
predominant_age_groups = gender_age_analysis[['Young Adults(18-35)', 'Middle Aged(36-60)', 'Seniors(61 and above)']].idxmax(axis=1)

# Display the results
print("Customer Gender vs. Age Group Analysis:")
print("\nDetailed Counts and Percentages:")
print("-------------------------")
print(gender_age_analysis[['Young Adults(18-35)', 'Middle Aged(36-60)', 'Seniors(61 and above)', 
                          'Total_Customers', 'Percent_of_Total',
                          'Young Adults(18-35) (%)', 'Middle Aged(36-60) (%)', 'Seniors(61 and above) (%)']])
print("\nSummary:")
print(f"Total Customers: {total_customers}")
print("\nPredominant age group by gender:")
for gender in predominant_age_groups.index:
    print(f"{gender}: {predominant_age_groups[gender]} "
          f"({gender_age_analysis.loc[gender, predominant_age_groups[gender]]} customers, "
          f"{gender_age_analysis.loc[gender, f'{predominant_age_groups[gender]} (%)']}%)")

Customer Gender vs. Age Group Analysis:

Detailed Counts and Percentages:
-------------------------
age_group  Young Adults(18-35)  Middle Aged(36-60)  Seniors(61 and above)  \
Gender                                                                      
Female                    1832                2498                    213   
Male                      2321                2885                    251   

age_group  Total_Customers  Percent_of_Total  Young Adults(18-35) (%)  \
Gender                                                                  
Female                4543             45.43                    40.33   
Male                  5457             54.57                    42.53   

age_group  Middle Aged(36-60) (%)  Seniors(61 and above) (%)  
Gender                                                        
Female                      54.99                       4.69  
Male                        52.87                       4.60  

Summary:
Total Customers: 10000

Predominant 

### Initial tentative observation

There are slightly more Middle aged female customers `55%` to male to `53%` while subsequesntly `42%` of Male customers are generally younger overall than the females `40%`

The seniors customers of both genders are on par at `5%` each

These further supports the insights indicating female customers who are middle aged to be at the most risk of churning.


In [27]:
# Customer Geography vs. Churn Analysis

# Grouping data by geography and Exited status
geography_churn_analysis = df.groupby(['Geography', 'Exited']).size().unstack(fill_value=0)

# Adding total customers column
geography_churn_analysis['Total_Customers'] = geography_churn_analysis['Churned'] + geography_churn_analysis['Retained']

# Calculate churn and retention rates for each geography
geography_churn_analysis['Churn_Rate'] = (geography_churn_analysis['Churned'] / geography_churn_analysis['Total_Customers'] * 100).round(2)
geography_churn_analysis['Retention_Rate'] = (geography_churn_analysis['Retained'] / geography_churn_analysis['Total_Customers'] * 100).round(2)

# Calculate percentage of total customers in each geography
total_customers = geography_churn_analysis['Total_Customers'].sum()
geography_churn_analysis['Percent_of_Total'] = (geography_churn_analysis['Total_Customers'] / total_customers * 100).round(2)

# Identify the geography with highest churn and retention rates
highest_churn_geography = geography_churn_analysis['Churn_Rate'].idxmax()
highest_retention_geography = geography_churn_analysis['Retention_Rate'].idxmax()

# Display the results
print("Customer Geography vs. Churn Analysis:")
print("\nDetailed Counts and Rates:")
print("-------------------------")
print(geography_churn_analysis[['Churned', 'Retained', 'Total_Customers', 'Percent_of_Total', 'Churn_Rate', 'Retention_Rate']])
print("\nSummary:")
print(f"Total Customers: {total_customers}")
print(f"Geography with highest churn rate: {highest_churn_geography} ({geography_churn_analysis.loc[highest_churn_geography, 'Churn_Rate']}%)")
print(f"Geography with highest retention rate: {highest_retention_geography} ({geography_churn_analysis.loc[highest_retention_geography, 'Retention_Rate']}%)")


Customer Geography vs. Churn Analysis:

Detailed Counts and Rates:
-------------------------
Exited     Churned  Retained  Total_Customers  Percent_of_Total  Churn_Rate  \
Geography                                                                     
France         810      4204             5014             50.14       16.15   
Germany        814      1695             2509             25.09       32.44   
Spain          413      2064             2477             24.77       16.67   

Exited     Retention_Rate  
Geography                  
France              83.85  
Germany             67.56  
Spain               83.33  

Summary:
Total Customers: 10000
Geography with highest churn rate: Germany (32.44%)
Geography with highest retention rate: France (83.85%)


### Initial tentative observations

`50%`of the customer base is located in France making it a key advertisement and product presence location.

Germany and spain follow suit at aprox `25%` each 

But surprisingly Germanhy has the highest percentage customer churn rate at `32%` leagues ahead of France and Spain who have rates of `16% and 17%` respectivley

More attention needs to paid to German customers to understand the factors contributting to the high churn rate there

In [28]:
# Customer Gender vs. Geography Analysis

# Grouping data by gender and geography
gender_geography_analysis = df.groupby(['Gender', 'Geography']).size().unstack(fill_value=0)

# Adding total customers column
gender_geography_analysis['Total_Customers'] = gender_geography_analysis.sum(axis=1)

# Calculate percentage distribution of geography within each gender
for location in ['France', 'Germany', 'Spain']:
    gender_geography_analysis[f'{location} (%)'] = (gender_geography_analysis[location] / gender_geography_analysis['Total_Customers'] * 100).round(2)

# Calculate percentage of total customers by gender
total_customers = gender_geography_analysis['Total_Customers'].sum()
gender_geography_analysis['Percent_of_Total'] = (gender_geography_analysis['Total_Customers'] / total_customers * 100).round(2)

# Identify the predominant geography for each gender
predominant_locations = gender_geography_analysis[['France', 'Germany', 'Spain']].idxmax(axis=1)

# Display the results
print("Customer Gender vs. Geography Analysis:")
print("\nDetailed Counts and Percentages:")
print("-------------------------")
print(gender_geography_analysis[['France', 'Germany', 'Spain', 
                                'Total_Customers', 'Percent_of_Total',
                                'France (%)', 'Germany (%)', 'Spain (%)']])
print("\nSummary:")
print(f"Total Customers: {total_customers}")
print("\nPredominant location by gender:")
for gender in predominant_locations.index:
    print(f"{gender}: {predominant_locations[gender]} "
          f"({gender_geography_analysis.loc[gender, predominant_locations[gender]]} customers, "
          f"{gender_geography_analysis.loc[gender, f'{predominant_locations[gender]} (%)']}%)")



Customer Gender vs. Geography Analysis:

Detailed Counts and Percentages:
-------------------------
Geography  France  Germany  Spain  Total_Customers  Percent_of_Total  \
Gender                                                                 
Female       2261     1193   1089             4543             45.43   
Male         2753     1316   1388             5457             54.57   

Geography  France (%)  Germany (%)  Spain (%)  
Gender                                         
Female          49.77        26.26      23.97  
Male            50.45        24.12      25.44  

Summary:
Total Customers: 10000

Predominant location by gender:
Female: France (2261 customers, 49.77%)
Male: France (2753 customers, 50.45%)


### Initial tentative observation

Unsurprisingly a large portion of both `male` and `female` customers `(approx 50% respectively)` reside in France due to it being a key customer stronghold

On closer inspection `26%` of `female` customer reside in `Germany` as against `24%` of the `male`,Germany has been established to have the highest churn rate already.

This is yet another factor isolating high income middle aged women as being more likely to churn

Further and indepth analysis would be made via visualizing the data in Google Looker studio where factors like credit score, tenure would be analyzed for its potential impact on churn

In [29]:
# Save the cleaned DataFrame to Excel
output_path = r"C:\Users\chris\OneDrive\Documents\paschal\MY PORTFOLIO PROJECTS\CUSTOMER CHURN\Churn_Modelling clean.xlsx"

# Save to Excel
df.to_excel(output_path, index=False)

print(f"File has been saved successfully to: {output_path}")

File has been saved successfully to: C:\Users\chris\OneDrive\Documents\paschal\MY PORTFOLIO PROJECTS\CUSTOMER CHURN\Churn_Modelling clean.xlsx


In [1]:

# Grouping data by exited status and salary categories
# Grouping data by gender and exited status
# Grouping gender by salary category 
# grouping gender and age
# grouping exited status by location 
# grouping gender by location
# grouping salary by location
# Grouping data by credit score status and exited
